[mlir][bufferization] Fix OneShotBufferize when `defaultMemorySpaceFn` is used

christopherbate opened the pull request:
As described in issue llvm/llvm-project#91518, a previous PR
llvm/llvm-project#78484 introduced the `defaultMemorySpaceFn` into
bufferization options, allowing one to inform OneShotBufferize that it
should use a specified function to derive the memory space attribute
from the encoding attribute attached to tensor types.

However, introducing this feature exposed unhandled edge cases,
examples of which are introduced by this change in the new test under
`test/Dialect/Bufferization/Transforms/one-shot-bufferize-encodings.mlir`.

Fixing the inconsistencies introduced by `defaultMemorySpaceFn` is pretty
simple. This change:

- Updates the `bufferization.to_memref` and `bufferization.to_tensor`
  operations to explicitly include operand and destination types,
  whereas previously they relied on type inference to deduce the
  tensor types. Since the type inference cannot recover the correct
  tensor encoding/memory space, the operand and result types must be
  explicitly included. This is a small assembly format change, but it
  touches a large number of test files.

- Makes minor updates to other bufferization functions to handle the
  changes in building the above ops.

- Updates bufferization of `tensor.from_elements` to handle memory
  space.


Integration/upgrade guide:

In downstream projects, if you have tests or MLIR files that explicitly use
`bufferization.to_tensor` or `bufferization.to_memref`, then update
them to the new assembly format as follows:

```
%1 = bufferization.to_memref %0 : memref<10xf32>
%2 = bufferization.to_tensor %1 : memref<10xf32>
```

becomes

```
%1 = bufferization.to_memref %0 : tensor<10xf32> to memref<10xf32>
%2 = bufferization.to_tensor %0 : memref<10xf32> to tensor<10xf32> 
```

christopherbate commented:
@matthias-springer Can you review? The vast majority of changes here are just to update tests due to change in assembly formats for `bufferization.to_tensor|to_memref`.

matthias-springer commented on mlir/include/mlir/Dialect/Bufferization/IR/BufferizationOps.td:
```diff
@@ -496,7 +501,8 @@ def Bufferization_ToMemrefOp : Bufferization_Op<"to_memref", [
                    "memref", "tensor",
-                   "memref::getTensorTypeFromMemRefType($_self)">
+                   "memref::getTensorTypeFromMemRefType($_self)",
+                   "bufferization::detail::tensorTypesMatchUpToEncoding">
```
I think you can use `AllShapesMatch` and `AllElementTypesMatch` instead of `tensorTypesMatchUpToEncoding`.

christopherbate replied:
Fixed

matthias-springer commented on mlir/lib/Dialect/Bufferization/Transforms/Bufferize.cpp:
```diff
@@ -67,10 +67,14 @@ BufferizeTypeConverter::BufferizeTypeConverter() {
     assert(inputs.size() == 1 && "expected exactly one input");
 
+    // Unranked to ranked casts must be explicit.
+    if (auto inputType = dyn_cast<UnrankedMemRefType>(inputs[0].getType()))
```
Why is this needed?

christopherbate replied:
I removed this. At the end of this materialization function there is ` llvm_unreachable("only tensor/memref input types supported");`, but if the input is `UnrankedMemRefType`, then it will still hit the `unreachable`.  Is that by design?

matthias-springer commented on mlir/test/Dialect/Bufferization/Transforms/one-shot-bufferize-encodings.mlir:
```diff
@@ -0,0 +1,133 @@
+// RUN: mlir-opt %s -one-shot-bufferize="use-encoding-for-memory-space" -split-input-file | FileCheck %s
+
+// TODO: move to tensor dialect tests
```
Can you address this TODO?

matthias-springer commented on mlir/test/Dialect/Bufferization/Transforms/one-shot-bufferize-encodings.mlir:
```diff
@@ -0,0 +1,133 @@
+// -----
+
+func.func @alloc_tesor_copy_from_non_default_space(%arg0: tensor<128xf32, 1>) -> tensor<128xf32, 2> {
+  %0 = bufferization.alloc_tensor() copy(%arg0) {memory_space = 2 : i64} : tensor<128xf32, 1>
```
What is the meaning of this op? The two memory spaces here are inconsistent.

christopherbate replied:
Since having a `copy` generates a `memref.copy`, this meaning is to copy data between memory spaces, however `bufferization.alloc_tensor` only prints one type for both `copy` operand and the result. So we will need to update ASM format for `bufferization.alloc_tensor` as well. That will remove the need for the `tensor.cast`.

matthias-springer commented on mlir/include/mlir/Dialect/Bufferization/IR/BufferizationOps.td:
```diff
@@ -388,9 +388,7 @@ def Bufferization_ToTensorOp : Bufferization_Op<"to_tensor", [
+    AllElementTypesMatch<["memref", "result"]>
   ]> {
   let summary = "create a tensor from a `memref`";
   let description = [{
```
Code examples should be updated in `description`.

christopherbate replied:
done

matthias-springer reviewed (commented):
Can you add extra unit tests for `scf.for` (with tensor iter_arg that has an encoding), `scf.if` and `scf.execute_region` (with tensor result). I think these bufferization pattern must be updated because they create new `to_tensor` ops.

matthias-springer commented:
> Can you add extra unit tests for `scf.for` (with tensor iter_arg that has an encoding), `scf.if` and `scf.execute_region` (with tensor result). I think these bufferization pattern must be updated because they create new `to_tensor` ops.

Ideally, to make this PR not even larger, if an update to these patterns is indeed needed, you could update those patterns in a first, separate PR. I think not much has to change, you just have to take the tensor types from the input IR instead of inferring them in the op builder.

aartbik commented:
lgtm for all sparse test signature changes

christopherbate committed:
[mlir][bufferization] Fix OneShotBufferize when `defaultMemorySpaceFn` is used

As described in issue llvm/llvm-project#91518, a previous PR
llvm/llvm-project#78484 introduced the `defaultMemorySpaceFn` into bufferization
options, allowing one to inform OneShotBufferize that it should use a specified
function to derive the memory space attribute from the encoding attribute
attached to tensor types.

However, introducing this feature exposed unhandled edge cases, examples of
which are introduced by this change in the new test under
`test/Dialect/Bufferization/Transforms/one-shot-bufferize-encodings.mlir`.

Fixing the inconsistencies introduced by `defaultMemorySpaceFn` is pretty
simple. This change:

- Updates the `bufferization.to_memref` and `bufferization.to_tensor` operations
  to explicitly include operand and destination types, whereas previously they
  relied on type inference to deduce the tensor types. Since the type
  inference cannot recover the correct tensor encoding/memory space, the
  operand and result types must be explicitly included. This is a small assembly
  format change, but it touches a large number of test files.

- Makes minor updates to other bufferization functions to handle the
  changes in building the above ops.

- Updates bufferization of `tensor.from_elements` to handle memory space.

matthias-springer commented on mlir/include/mlir/Dialect/Bufferization/IR/Bufferization.h:
```diff
@@ -12,10 +12,16 @@
+#include "llvm/Support/Debug.h"
+
+namespace mlir::bufferization::detail {
+bool tensorTypesMatchUpToEncoding(Type lhs, Type rhs);
```
This is not used.

matthias-springer commented on mlir/lib/Dialect/Bufferization/Transforms/Bufferize.cpp:
```diff
@@ -222,6 +222,13 @@ struct OneShotBufferizePass
             [](TensorType t) -> std::optional<Attribute> {
           return std::nullopt;
         };
+      } else if (useEncodingForMemorySpace) {
```
Can you emit an error/assertion if both `mustInferMemorySpace` and `useEncodingForMemorySpace` are set?

srcarroll commented:
is this PR dead? wondering why it's been sitting here approved for months

christopherbate commented:
It's not dead; I have just been pressed for time. Let me rebase and address Matthias' last comments this week.

CoTinker commented:
Hi, this PR is useful for me, do you have any time to rebase it. Thanks.
